### PR TITLE
External storage 'Login credentials' auth mechanism

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
This auth mechanism saves user credentials in the database, encrypted. This is done with the new CredentialsManager class.

:warning: The `setValues()` method on `IDBConnection` might break with Oracle. Hopefully the unit tests will pass, but if not, the fix lies in `to_char(...)` in the WHERE clause when using clobs. :warning:

Please test and review @PVince81 @MorrisJobke @DeepDiver1975 @icewind1991 @LukasReschke 

@MTRichards ref #16305

### CODE REVIEW

**Comment by PVince81** [comment]
Please rebase

File: apps/files_external/appinfo/application.php [hunk 1]
```diff
@@ -91,6 +91,7 @@ protected function loadAuthMechanisms() {
 			// AuthMechanism::SCHEME_PASSWORD mechanisms
 			$container->query('OCA\Files_External\Lib\Auth\Password\Password'),
 			$container->query('OCA\Files_External\Lib\Auth\Password\SessionCredentials'),
+			$container->query('OCA\Files_External\Lib\Auth\Password\LoginCredentials'),
```
Comment (icewind1991) [inline apps/files_external/appinfo/application.php]: needs a beter name

File: apps/files_external/lib/auth/password/logincredentials.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		\OCP\Util::connectHook('OC_User', 'post_login', $this, 'authenticate');
+	}
+
+	/**
+	 * Hook listener on post login
+	 *
+	 * @param array $params
+	 */
+	public function authenticate(array $params) {
+		$user = $params['uid'];
```
Comment (icewind1991) [inline apps/files_external/lib/auth/password/logincredentials.php]: call this variable `$userId` to reduce confusion with the loginname

File: lib/private/db/connection.php [hunk 3]
```diff
@@ -243,6 +244,52 @@ public function insertIfNotExist($table, $input, array $compare = null) {
...
+	 * @param string $table
+	 * @param array $keys (column name => value)
+	 * @param array $values (column name => value)
+	 * @param array $updatePreconditionValues ensure values match preconditions (column name => value)
+	 * @return int number of new rows
+	 * @throws \Doctrine\DBAL\DBALException
+	 * @throws PreconditionNotMetException
+	 */
+	public function setValues($table, array $keys, array $values, array $updatePreconditionValues = []) {
+		try {
```
Comment (icewind1991) [inline lib/private/db/connection.php]: Can lead to race conditions.

We have insertifnotexists for this

Comment (Xenopathic) [inline lib/private/db/connection.php]: This function tries an insert first, then tries an update. Using `insertIfNotExists()` instead of the current insert would have exactly the same effect - we still need to try an update afterwards if the insert fails.

**Comment by DeepDiver1975** [comment]
@PVince81 @Xenopathic schedule for 9.0

File: db_structure.xml [hunk 4]
```diff
@@ -1192,5 +1192,60 @@
...
+			<field>
+				<name>user</name>
+				<type>text</type>
+				<default></default>
+				<notnull>false</notnull>
+				<length>64</length>
+			</field>
+
+			<field>
+				<name>identifier</name>
```
Comment (PVince81) [inline db_structure.xml]: add comment, what is this ?

Comment (PVince81) [inline db_structure.xml]: does it link to the mount id ?

**Comment by PVince81** [comment]
Please rebase :smile:

**Comment by Xenopathic** [comment]
I'd rather we wait for the config to live in the DB fully before merging this, for cleaner, better code.

**Comment by PVince81** [comment]
@Xenopathic the config is already in the DB. Please rebase :smile: 

@icewind1991 please notice this PR about auth mechanism, to make sure we're not duplicating work :smile:

**Comment by icewind1991** [comment]
rebased and added more tests for `setValues`

**Comment by icewind1991** [comment]
Fixed checking the external storage status (`owncloud/index.php/apps/files_external/userglobalstorages/$id`) by passing the user when manipulating storage config

**Comment by icewind1991** [comment]
After applying #21467 everything seems to work fine for an admin defined mount using login credentials.

Storage is accessible both trough the WI/DAV and occ

**Comment by Xenopathic** [comment]
I still don't like the approach of the `CredentialsManager` in core, ideally the configuration row in the files_external tables should be updated/created, but that approach isn't so simple with multiple users access to a global storage.

I had an idea (completely conceptual at this stage) to create a wrapping storage per-user that contains the credentials, along with a 'pointer' to the global storage. Then during mounting, that wrapped storage overrides the global storage and can apply the credentials as necessary. That would also form a nice basis for the 'custom credentials' authentication mechanism (as required by the WND merger) which would work in a similar way, but with those custom credentials directly modifiable in the configuration UI.

**Comment by icewind1991** [comment]
I think the current idea is better and less complex.

The CredentialsManager can also be used to handle the custom credentials

**Comment by PVince81** [comment]
Doesn't seem to work. Maybe I tested it wrong:

1. Setup a new SFTP or SMB ext storage
1. Select "Login credentials"
1. Enter usual settings
1. Leave applicable to "All"
1. Create an OC user using login + password of the storage
1. Login as that user

Expected: storage appears
Actual: storage folder does not appear, notification says that storage not configured properly and click on the red row, but there is no row at all. Personal page also contains no info.

@icewind1991 @Xenopathic how did you test ?

File: apps/files_external/controller/userglobalstoragescontroller.php [hunk 5]
```diff
@@ -78,12 +88,22 @@ public function index() {
...
+		$backend->manipulateStorageConfig($storage, $this->userSession->getUser());
+	}
+
 	/**
 	 * Get an external storage entry.
 	 *
 	 * @param int $id storage id
 	 * @return DataResponse
 	 *
+	 * @NoCSRFRequired
```
Comment (PVince81) [inline apps/files_external/controller/userglobalstoragescontroller.php]: intended?

File: apps/files_external/lib/auth/password/logincredentials.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+	/**
+	 * Hook listener on post login
+	 *
+	 * @param array $params
+	 */
+	public function authenticate(array $params) {
+		$userId = $params['uid'];
+		$credentials = [
+			'user' => $this->session->get('loginname'),
```
Comment (PVince81) [inline apps/files_external/lib/auth/password/logincredentials.php]: Assuming that this is correct, will this work with LDAP user ids ?

Comment (PVince81) [inline apps/files_external/lib/auth/password/logincredentials.php]: CC @blizzz

**Comment by PVince81** [comment]
Please increase the app's version. Seems that when I switched to this branch it did not create the table.

Will retest from scratch with a new install.

**Comment by PVince81** [comment]
Okay, it works from scratch.

@icewind1991 please increase the app version

File: apps/files_external/controller/userglobalstoragescontroller.php (hunk 5 above)
Comment (icewind1991) [inline apps/files_external/controller/userglobalstoragescontroller.php]: leftover from debuging :speak_no_evil:

**Comment by icewind1991** [comment]
> please increase the app version

Done

**Comment by PVince81** [comment]
:+1:

Maybe @Xenopathic and @icewind1991 could review each other's code ?

Unless @jmaciasportela wants to chime in for the final review ?

**Comment by icewind1991** [comment]
:+1: for @Xenopathic's part

File: apps/files_external/lib/auth/password/logincredentials.php (hunk 6 above)
Comment (jmaciasportela) [inline apps/files_external/lib/auth/password/logincredentials.php]: Historical problem in WND + LDAP https://github.com/owncloud/windows_network_drive/issues/218#issuecomment-108363607

**Comment by icewind1991** [comment]
> 06:44:45 There was 1 error:
06:44:45 
06:44:45 1) Test\TestAllConfig::testSetUserValueWithPreCondition
06:44:45 Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'UPDATE "oc_preferences" SET "configvalue" = ? WHERE ("userid" = ?) AND ("appid" = ?) AND ("configkey" = ?) AND ("configvalue" = ?)' with params ["valuePreCond2", "userPreCond", "appPreCond", "keyPreCond", "valuePreCond"]:
06:44:45 
06:44:45 ORA-00932: inconsistent datatypes: expected - got CLOB

Will look into it

(jenkins-oci actually reporting real failures, amazing)

**Comment by Xenopathic** [comment]
Yup, as predicted in the original PR post, Oracle broke. @nickvergessen had a PR a while back that simply removes the WHERE clause on Oracle, making things slower sure, but at least not breaking.

**Comment by nickvergessen** [comment]
clob fields can not be compared on OCI

**Comment by icewind1991** [comment]
Fixed by adding adding support for comparing against clobs on oci if you explicitly define you're trying to compare strings

cc @DeepDiver1975 for the query builder stuff (last commit)

**Comment by icewind1991** [comment]
all green, can this be merged?

**Comment by PVince81** [comment]
:+1: from me for the last changes

@nickvergessen @DeepDiver1975 is the query builder change acceptable ? https://github.com/owncloud/core/commit/58afddfaa585fdb9efb34c01d1a5fa6282ed2bd1

File: lib/public/db/querybuilder/iexpressionbuilder.php [hunk 7]
```diff
@@ -84,11 +84,13 @@ public function comparison($x, $operator, $y);
...
 	 * @param mixed $x The left expression.
 	 * @param mixed $y The right expression.
+	 * @param int|null $type @since 9.0.0 one of the \PDO::PARAM_* constants
+	 *                  required when comparing text fields for oci compatibility.
 	 *
 	 * @return string
 	 * @since 8.2.0
 	 */
-	public function eq($x, $y);
+	public function eq($x, $y, $type = null);
```
Comment (nickvergessen) [inline lib/public/db/querybuilder/iexpressionbuilder.php]: I think if we do this, we should do it for all comparisons, or at least `neq()`

Should also then add unit tests for this directly.
Can also do this in another PR, up to @DeepDiver1975

Comment (nickvergessen) [inline lib/public/db/querybuilder/iexpressionbuilder.php]: In order to keep the methods from doctrine untouched, we could also add a new method, that would make the comparison explicit, while not touching the interface of doctrine methods

**Comment by PVince81** [comment]
> In order to keep the methods from doctrine untouched, we could also add a new method, that would make the comparison explicit, while not touching the interface of doctrine methods

@icewind1991 can you make the change + add unit tests ? (or make it a separate PR so we can review it separately)

File: lib/public/db/querybuilder/iexpressionbuilder.php (hunk 7 above)
Comment (icewind1991) [inline lib/public/db/querybuilder/iexpressionbuilder.php]: I would prefer to keep it as arguments to the existing methods

**Comment by icewind1991** [comment]
> In order to keep the methods from doctrine untouched, we could also add a new method, that would make the comparison explicit, while not touching the interface of doctrine methods

I prefer to merge this as is and we can discuss whether to make it a new method or keep it as argument in a seperate issue/pr.
It's a shame to keep this PR blocked over a minor issue like that

cc @nickvergessen

File: lib/public/db/querybuilder/iexpressionbuilder.php (hunk 7 above)
Comment (PVince81) [inline lib/public/db/querybuilder/iexpressionbuilder.php]: Can you add `neq` then too and the unit test for both methods, if possible ?

Comment (nickvergessen) [inline lib/public/db/querybuilder/iexpressionbuilder.php]: Well we can do that once we decide on the API and how we want to handle it. Until then lets waste as few effort as possible

**Comment by PVince81** [comment]
@jmaciasportela @DeepDiver1975 can you help with the second review ?
or @Xenopathic review @icewind1991's latest changes so we can merge this

**Comment by jmaciasportela** [comment]
Except this comment only to take into account: https://github.com/owncloud/core/pull/18531/files#r49991237
LGTM :+1:

File: apps/files_external/lib/auth/password/logincredentials.php (hunk 6 above)
Comment (PVince81) [inline apps/files_external/lib/auth/password/logincredentials.php]: I'll add a note in the original ticket for QA to test this case
